### ISSUE TITLE
memory leaks when transmitting data in binary format

### ISSUE THREAD

**Issue by webarchymeta** [node-webrtc/node-webrtc#304]
It was found that when sending/receiving data in binary format (namely the Uint8Array format, see #303), some of the memory allocated is never released under Linux and Mac OS X, resulting in a continuously growing memory size used by the application. For Windows, the memory used can be kept at a constant, but relatively large value.

However, when encoding the data using base64 encoding, the memory used by the same application are kept at a quite smaller stable value ...

**Comment by SourceBoy** [comment]
I can confirm this as well. When transferring a file, both sender & recipient's resident memory sizes increase as the file transfers, and they're not released after transfer is complete.

Given a file that's 271.8MB, when the transfer is complete, the resident memory on both ends are roughly around the size of the file (271.8MB):
```
base64 (ok):
        sender
                start { rss: 55734272, heapTotal: 32542720, heapUsed: 18060600 }
                end { rss: 71114752, heapTotal: 26251264, heapUsed: 15908624 }
        recipient
                start { rss: 56528896, heapTotal: 32542720, heapUsed: 17995136 }
                end { rss: 53059584, heapTotal: 23105536, heapUsed: 16675608 }

Uint8Array (memory issue):
        sender
                start mem { rss: 55701504, heapTotal: 32542720, heapUsed: 18082848 }
                end mem { rss: 364601344, heapTotal: 57716736, heapUsed: 29253696 }
        recipient
                start mem { rss: 56270848, heapTotal: 32542720, heapUsed: 18007416 }
                end mem { rss: 321478656, heapTotal: 22056960, heapUsed: 18232512 }
```

I've also ruled out the possibilities that this could've been caused by `Uint8Array` or Node's `stream` with separate tests using the same file with `Uint8Array` but without `RTCDataChannel.send()`:
```
Uint8Array (ok)
	sender
		start { rss: 27607040, heapTotal: 10522624, heapUsed: 4499552 }
		end { rss: 69107712, heapTotal: 49680384, heapUsed: 26788840 }
	recipient
		start { rss: 27291648, heapTotal: 10522624, heapUsed: 4385976 }
		end { rss: 40894464, heapTotal: 14716928, heapUsed: 5166192 }
```

I've previously filed a similar issue in [#205](https://github.com/js-platform/node-webrtc/issues/205#issuecomment-274725722)

**Comment by vikeshkhanna** [comment]
+1

I was also able to isolate the problem to `dataChannel.send()` when using binaryType='arraybuffer'. The memory does not build up when sending base64 encoded strings.

Any update on this? Any pointers?

**Comment by cciliber** [comment]
I might have found a (the?) problem. 

At [Line 249 in datachannel.cc](https://github.com/js-platform/node-webrtc/blob/23a93521a7f7f76d920a7e6b27b3511e1106fda1/src/datachannel.cc#L249) there is a call to 

```c
arraybuffer->Externalize();
```

But from the [V8 documentation](https://v8docs.nodesource.com/node-5.12/d5/d6e/classv8_1_1_array_buffer.html#a9291f6ac203b9ceae83f7f17d39ecb59), `Externalize()`:

> Make this ArrayBuffer external. The pointer to underlying memory block and byte length are returned as |Contents| structure. After ArrayBuffer had been extrenalized, it does no longer owns the memory block. **The caller should take steps to free memory** when it is no longer needed.

However the corresponding memory is never freed (**Note** that the call to ```Neuter()``` at Line [265](https://github.com/js-platform/node-webrtc/blob/23a93521a7f7f76d920a7e6b27b3511e1106fda1/src/datachannel.cc#L265) does not free the memory). 

A solution that seems to work on my setup is to substitute 
```c
v8::ArrayBuffer::Contents content = arraybuffer->Externalize();
```
with
```c
v8::ArrayBuffer::Contents content = arraybuffer->GetContents();
```
which does not take ownership of the memory, and of course **remove** `arraybuffer->Neuter();` since non-externalized ArrayBuffers cannot be Neutered.

Hope this helps!

**Comment by webarchymeta** [comment]
I have tested using v0.0.66, the leak problem has not been solved ....

**Comment by markandrus** [comment]
@webarchymeta thanks for checking. I think there is a second problem yet to be addressed with the MessageEvent raised on the DataChannel. I’m AFK right now, but will check this weekend. BTW, can you share a test/technique for measuring the memory leak? Thanks, Mark

**Comment by webarchymeta** [comment]
We use the compiled version of wrtc in a data tunneling system, which is kind of complex to extract the required bits for testing. The problem was found when streaming HD video, the memory usage on the receiving end keeps growing (MessageEvent related ?) during the data transfer. I will keep testing newly released version of wrtc using the same system.

I'd like to work on the source level of wrtc, however the network environment I am in makes it really none-trivial to download anything from certain parts of the web (e.g. anything to do with chrome, etc.) ...

**Comment by nazar-pc** [comment]
Now since other issues I'm aware of were fixed (thanks @markandrus for that!) I found that there is likely a memory leak.

Consider following reduced test case:
```javascript
const wrtc     = require('wrtc');
const peer1    = new wrtc.RTCPeerConnection({iceServers : []});
const peer2    = new wrtc.RTCPeerConnection({iceServers : []});
const channel1 = peer1.createDataChannel('data');

channel1.onopen = function () {
    const string = '0'.repeat(10000);
    const array  = new Uint8Array(10000);
    setInterval(function () {
        // channel1.send(string);
        channel1.send(array);
    });
};

peer1.onicecandidate = function (event) {
    if (event.candidate) {
        peer2.addIceCandidate(event.candidate);
    }
};

peer2.onicecandidate = function (event) {
    if (event.candidate) {
        peer1.addIceCandidate(event.candidate);
    }
};

peer1.createOffer()
    .then(function (offer) {
        return peer1.setLocalDescription(offer);
    })
    .then(function () {
        peer2.setRemoteDescription(peer1.localDescription);
        return peer2.createAnswer();
    })
    .then(function (answer) {
        return peer2.setLocalDescription(answer);
    })
    .then(function () {
        return peer1.setRemoteDescription(peer2.localDescription);
    });
```

When using 10k `Uint8Array` memory goes up relatively quickly. With string it also grows constantly, but not nearly as fast.

**Comment by markandrus** [comment]
Nice, thank you for the test case. How do you measure the memory usage? Also, I think it may make sense to augment the test to expose `gc()` and to try calling it at certain moments to see if it may affect memory usage.

**Comment by nazar-pc** [comment]
I look at GNOME System Monitor. I've tried forcing garbage collection using Chromium Dev Tools, doesn't affect anything at all.

I suspect it might be somehow related to underlying libwebrtc, since I see very similar behavior with latest Chromium Nightly (the only difference is that it refuses to send any data after some short time): https://bugs.chromium.org/p/webrtc/issues/detail?id=9098

**Comment by markandrus** [comment]
OK, I believe the issue is because we don't delete any of the events we enqueue in the RTCPeerConnection and RTCDataChannel `Run` functions.

**Comment by nazar-pc** [comment]
After an hour of running above script with #387 didn't notice memory leaking anymore 🚀

**Comment by markandrus** [comment]
Excellent. I'll finish the other bullet list item and then aim for a 0.1.2 release.

**Comment by markandrus** [comment]
The biggest leak on RTCDataChannel should be fixed now in v0.1.2.
